src/memoize/memoize.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2014-present Palantir Technologies
+ * @license MIT
+ * @fileoverview Implements a function memoizer using the Signature API.
+ */
+
+import { signArray, Signature } from "./signature";
+
+export interface IMemoizedFunctionMethods {
+    /**
+     * Lock the memoization to always return the cached
+     * property during the duration of fn. This lets you
+     * bypass the performance hit of signing when you
+     * know the fn will not mutate the inputs.
+     *
+     * Be sure to force the memoization to the value
+     * you want before calling this!
+     *
+     * @param fn
+     */
+    doLocked<T>(fn: () => T): T;
+
+    /**
+     * Turn on perf logging. Useful for debugging.
+     */
+    logPerformance(): this;
+}
+
+export type MemoizedFunction<F extends Function> = F & IMemoizedFunctionMethods;
+
+export function memoize<F extends Function>(compute: F): MemoizedFunction<F> {
```

Comment:
alright, so this is the actual memoize function that will be the main first thing people read. i'd like the comment to tell me:

how to use it (short code example).
what sort of pitfalls there are (vis a vis signatures). a warning that arguments that are not shallow comparable may need a custom signature.
also,  what memoization locking is and why you would want to use it.